torch/autograd/gradcheck.py

```diff
@@ -661,6 +762,223 @@ def transpose(matrix_of_tensors):
+            if a.numel() != 0 or n.numel() != 0:
+                if o.is_complex():    # C -> C, R -> C
+                    if not torch.allclose(analytical_from_imag_grad_out[j], numerical_from_imag_grad_out[i][j], rtol, atol):
+                        return fail_test(get_notallclose_msg(analytical_from_imag_grad_out[j],
+                                                             numerical_from_imag_grad_out[i][j], i, j,
+                                                             "Gradients failed to compare equal for grad output = 1j. "))
+                if inp.is_complex():  # C -> R, C -> C
+                    if not torch.allclose(a, n, rtol, atol):
+                        return fail_test(get_notallclose_msg(a, n, i, j,
+                                                             "Gradients failed to compare equal for grad output = 1. "))
+                else:                 # R -> R, R -> C
+                    if not torch.allclose(a, n, rtol, atol):
+                        return fail_test(get_notallclose_msg(a, n, i, j))
+    return True
+
+
+def dot(u, v):
+    if v.is_complex() and not u.is_complex():
+        return v.dot(u.to(dtype=v.dtype))
+    elif u.is_complex() and not v.is_complex():
+        return u.dot(v.to(dtype=u.dtype))
+    elif u.is_complex() and v.is_complex():
+        return u.to(dtype=torch.complex128).dot(v.to(dtype=torch.complex128))
+    else:
+        return u.to(dtype=torch.float64).dot(v.to(dtype=torch.float64))
+
+
+def all_close(a, b, rtol, atol):
+    if a.is_complex() and not b.is_complex():
+        if not torch.allclose(a, b.to(a.dtype), rtol, atol):
+            return False
+    elif b.is_complex() and not a.is_complex():
+        if not torch.allclose(a.to(b.dtype), b, rtol, atol):
+            return False
+    elif a.is_complex() and b.is_complex():
+        if not torch.allclose(a.to(dtype=torch.complex128), b.to(dtype=torch.complex128), rtol, atol):
+            return False
+    else:
+        if not torch.allclose(a.to(dtype=torch.float64), b.to(dtype=torch.float64), rtol, atol):
+            return False
+    return True
+
+
+def vec_from_tensor(x, generator):
+    # Create a random vector with the same number of elements as x and the same dtype/device
+    # If x is complex, we create a complex tensor with only real component
+    if x.layout == torch.sparse_coo:
+        # For sparse, create a random sparse vec with random values in the same
+        # indices. Make sure size is set so that it isn't inferred to be smaller.
+        x_values = x._values()
+        values = torch.rand(x_values.numel(), generator=generator) \
+            .to(dtype=x.dtype, device=x.device) \
+            .reshape(x_values.shape)
+        values /= values.norm()
+        vec = torch.sparse_coo_tensor(x._indices(), values, x.size())
+    else:
+        vec = torch.rand(x.numel(), generator=generator).to(dtype=x.dtype, device=x.device)
+        vec /= vec.norm()
+    return vec
```

Comment:
Why do we have to return a vector (1D tensor) for `vec_from_tensor`? (I noticed that this function still maintains this behavior in the refactor PR). In the ["minimal fast gradcheck example"](https://gist.github.com/zou3519/5f4e3b4f77655b469160574b4b8813e8) there's no need for the u and v vectors (tangents, cotangents) to be flat vectors

Reply:
If we have them as 1D vecs we can reuse the functions we have for the slow mode as-is because in slow-mode we treat them as cols of a jacobian matrix

Reply:
Is it cleaner if we don't reuse the functions for slow-mode? A lot of the slow mode code needs to do some careful bookkeeping (e.g. set individual elements to 1 for the analytical jacobian computation). Because fast mode doesn't need to do such careful bookkeeping we should be able to write some cleaner code if it goes through a different mechanism

Reply:
I tried using non-flat tensors for the tangents cotangents (while still reusing the slow code) its mainly adding a lot of special cases to handle `if fast, don't flatten; else flatten` mainly in the analytical case.

I don't know how much it would simplify if we try to do a complete rewrite, but the part where we do the careful bookkeeping of onehot vectors should already be factored out into `compute_analytical_jacobian_rows`.

Reusing the functions for slow here mainly allows us to avoid writing error checking (size/dtype/reentrancy), then printing out the error message.

In the numerical case there is already very little code reuse (proportionally). The main complexity of handling sparse/opaque tensors needs to be done is slightly different ways (due to one-hot iteration). Its mainly the logic that handles complex inputs/outputs computes the gradients that we reuse.

Reply:
sounds good, thanks for thinking through this